src/main/java/io/confluent/kafkarest/ConsumerReadTask.java

```diff
@@ -108,37 +110,50 @@ public boolean doPartialRead() {
       long startedIteration = parent.getConfig().getTime().milliseconds();
       final int requestTimeoutMs = parent.getConfig().getInt(
           KafkaRestConfig.CONSUMER_REQUEST_TIMEOUT_MS_CONFIG);
+      final long endTime = startedIteration + requestTimeoutMs;
+      final int itBackoff = parent.getConfig().getInt(
+        KafkaRestConfig.CONSUMER_ITERATOR_BACKOFF_MS_CONFIG);
       try {
-        // Read off as many messages as we can without triggering a timeout exception. The
-        // consumer timeout should be set very small, so the expectation is that even in the
-        // worst case, num_messages * consumer_timeout << request_timeout, so it's safe to only
-        // check the elapsed time once this loop finishes.
-        while (iter.hasNext()) {
-          MessageAndMetadata<KafkaK, KafkaV> msg = iter.peek();
-          ConsumerRecordAndSize<ClientK, ClientV> recordAndSize = parent.createConsumerRecord(msg);
-          long roughMsgSize = recordAndSize.getSize();
-          if (bytesConsumed + roughMsgSize > maxResponseBytes) {
-            break;
+        while (parent.getConfig().getTime().milliseconds() < endTime) {
+
+          if (iter == null || !iter.hasNext()) {
+            // The consumer timeout should be set very small, so the expectation is that even in the
+            // worst case, num_messages * consumer_timeout << request_timeout, so it's safe to only
+            // check the elapsed time once this loop finishes.
+            ConsumerRecords<KafkaK, KafkaV> records = consumer.poll(itBackoff);
+            iter = records.iterator();
+            if (!iter.hasNext()) {
+              // there are no records left
+              backoff = true;
+              break;
+            }
           }
 
-          iter.next();
-          messages.add(recordAndSize.getRecord());
-          bytesConsumed += roughMsgSize;
-          // Updating the consumed offsets isn't done until we're actually going to return the
-          // data since we may encounter an error during a subsequent read, in which case we'll
-          // have to defer returning the data so we can return an HTTP error instead
+          while (iter.hasNext()) {
+            org.apache.kafka.clients.consumer.ConsumerRecord<KafkaK, KafkaV> record = iter.next();
```

Comment:
I'm actually sort of hoping it might be possible to delete the `ConsumerRecord` we have here and use the one in Kafka only. I think we would just need to add a custom Jackson module to ensure that it serializes correctly. I don't think we necessarily need to do it here though.